relay-sampling/src/lib.rs

```diff
@@ -809,79 +839,184 @@ impl Default for SamplingMode {
     }
 }
 
+/// Represents a list of rule ids which is used for outcomes.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct MatchedRuleIds(pub Vec<RuleId>);
+
+impl MatchedRuleIds {
+    /// Creates a MatchedRuleIds struct from a string formatted with the following format:
+    /// rule_id_1;rule_id_2;...
+    pub fn from_string(value: &str) -> Result<MatchedRuleIds, ParseIntError> {
+        let mut rule_ids = vec![];
+
+        for rule_id in value.split(';') {
+            let int_rule_id = rule_id.parse()?;
+            rule_ids.push(RuleId(int_rule_id));
+        }
+
+        Ok(MatchedRuleIds(rule_ids))
+    }
+}
+
+impl Display for MatchedRuleIds {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "{}",
+            self.0
+                .iter()
+                .map(|rule_id| format!("{rule_id}"))
+                .collect::<Vec<String>>()
+                .join(";")
+        )
+    }
+}
+
+/// Represents the specification for sampling an incoming event.
+#[derive(Clone, Debug, PartialEq)]
+pub struct SamplingConfigMatchResult {
+    pub sample_rate: f64,
+    pub seed: Uuid,
+    pub matched_rule_ids: MatchedRuleIds,
+}
+
 /// Represents the dynamic sampling configuration available to a project.
 ///
 /// Note: This comes from the organization data
 #[derive(Debug, Clone, Serialize, Deserialize)]
 #[serde(rename_all = "camelCase")]
 pub struct SamplingConfig {
-    /// The ordered sampling rules for the project from highest to lowest priority.
+    /// The ordered sampling rules for the project.
+    ///
+    /// This field will remain here to serve only for old customer Relays to which we will
+    /// forward the sampling config. The idea is that those Relays will get the old rules as
+    /// empty array, which will result in them not sampling and forwarding sampling decisions to
+    /// upstream Relays.
+    #[serde(default, skip_deserializing)]
     pub rules: Vec<SamplingRule>,
+    /// The ordered sampling rules v2 for the project.
+    pub rules_v2: Vec<SamplingRule>,
```

Comment:
This might make the protocol a bit confusing, but I think it makes the rest of the code / diff easier to read. Feel free to disagree.

```suggestion
    #[serde(default, skip_deserializing, rename="rules")]
    pub legacy_rules: Vec<SamplingRule>,
    /// The ordered sampling rules v2 for the project.
    #[serde(rename = "rulesV2")]
    pub rules: Vec<SamplingRule>,
```

Reply:
I would prefer to keep the `rules_v2` schema, to be more explicit about the newer version, like it is typically done in APIs. The reasoning for avoiding `legacy` is that if we upgrade again the schema, we will have to define `legacy_legacy`? I envision a simpler naming scheme with `rules_v*`. Ofc the best solution would be to implement proper versioning.